lib/runners/java.js

```diff
@@ -1,86 +1,197 @@
 var shovel = require('../shovel'),
-    exec = require('child_process').exec,
-    util = require('../util');
+    util = require('../util'),
+    tagHelpers = require('../utils/tag-helpers'),
+    manipulateFileSync = require('../utils/manipulate-file-sync'),
+    fs = require('fs-extra');
 
 module.exports.run = function run(opts, cb) {
+  // we will cache the contents of this file so that we can process it later.
+  let buildGradle = '';
 
   shovel.start(opts, cb, {
     solutionOnly: function(runCode, fail) {
+      fs.emptydirSync(`${opts.dir}/src`);
+
       const names = extractNames(opts.solution, 'Solution');
-      const solutionFile = util.codeWriteSync('javac', opts.solution, opts.dir, names.file);
-
-      compile([solutionFile], function(error, stdout, stderr) {
-        if (error) return fail(error, stdout, stderr);
-        opts.publish('stdout', stdout);
-        runCode({
-          name: 'groovyclient',
-          args: [
-            '-cp', opts.dir,
-            '-e', `import ${names.full}; ${names.full}.main(null);`,
-            '-Cauthtoken', 'groovy'
-          ]
-        });
-      });
+      util.codeWriteSync(null, opts.solution, `${opts.dir}/src/main/java`, names.file);
```

Comment:
Yeah I want to phase `writeFileSync` out. I can update this to not use it. 

I didn't initially add the codeWriteSync stuff, that came from someone else who used to contribute to the project, but the basic idea I guess was just to encapsulate the idea of having to name files dynamically based off of the contents of the code. I think this design also came around a time when there was more stuff in the config.js file (like shared code between CoffeeScript and JS), and people just felt compelled to use it more. I'm not sure if this really answers your question though.

The entire config.js needs a revamp. I don't know that it needs to really exist at all anymore actually. The utilities need a revamp too. I want to move each utility method into its own file and cleanup all of the usages, because right now writeFileSync and codeWriteSync are ambiguous and hard to reason about. 

I also want to clean shovel.js up because it sucks. In this PR I renamed `exec` to `spawnEx` to at least not make it completely silly (since `exec` is something different in Node). I would however like to rip most of this stuff out into their own utility files and clean things up a bit so that its easier to follow the logic. Also upgrading to Node 8 and using awaits would make things easier to follow.